README.rst
AWStatic is a front-end for `AWStats <http://awstats.sourceforge.net/>`_:
it reads AWStats data files and produces a static HTML file that can
be accessed offline or served by any HTTP server (without CGI
support).

The generated report heavily uses JavaScript and cannot be properly
displayed if JavaScript support is disabled in your browser.

For now, AWStatic is not much more than a small exercise in building a
client-side application. It is far from being as complete as the
default front-end shipped with AWStats (although it probably is a bit
more elegant). It currently includes:

- a full statistics overview with hits, pages, visits, visitors and
  bandwidth by day, by month and by year;

- top 10 pages (hits and bandwidth) by month and by year;

- top 10 files (hits and bandwidth) by month and by year;

- top 30 referrers (pages and hits) by month and by year;

- top 30 search keywords by month and by year;

- top 30 search phrases by month and by year.

A future version may include diagrams that show the repartition of
operating systems and browsers.


Screenshots
===========

.. image:: https://raw.github.com/dbaty/AWStatic/master/docs/overview.png
   :width: 518
   :height: 317
   :align: center
   :alt: The overview

.. image:: https://raw.github.com/dbaty/AWStatic/master/docs/top10.png
   :width: 361
   :height: 235
   :align: center
   :alt: The top 10 pages


Installation
============

As usual, installing in a virtual environment is highly recommended:

.. code-block:: bash

   $ easy_install https://github.com/dbaty/AWStatic/tarball/master


How to use AWStatic
===================

Once installed, AWStatic provides an ``awstatic`` executable:

.. code-block:: bash

   $ awstatic --help
   usage: awstatic [-h] [-v] [CONFIG_FILE]

   positional arguments:
     CONFIG_FILE    The configuration file to use. Default is "./awstatic.ini".

   optional arguments:
     -h, --help     show this help message and exit
     -v, --version  show program's version number and exit


As shown in the help message, AWStatic requires a configuration
file. Here is an example:

.. code-block:: ini

   [awstatic]
   awstats_dir = /var/lib/awstats
   out_dir = ./out
   sites = www.example.com=http://example.com
           www.anotherexample.com=http://another.example.com
   pdb = true

   [logger]
   level = debug
   path = -

There are two sections: ``awstatic`` and ``logger``. In the
``awstatic`` section, you may use the following options:

``awstats_dir``
    Directory where raw AWStats data reside. On Debian, it is
    ``/var/lib/awstats`` by default. If you are not sure where those
    files are located on your system, look for files named
    ``awstats.*.txt``.

``out_dir``
    Directory where AWStatic will write its report. 

``sites``
    A list of sites to include in the report. Sites are separated by
    spaces or newline characters. Each site is defined by an
    identifier that has been used to generate AWStats data files,
    followed by an equal sign (``=``), followed by the root URL of the
    site (which will be used by AWStatic to provide fully qualified
    URLs in the report).

    If multiple sites are provided, the report will include data for
    all sites and the user will have to select the site she is
    interested in (see the screenshot above). Sites are supposed to be
    independant and their data are **not** merged.

``pdb``
    A debugging option, useful only if you feel adventurous and would
    like to jump in the code when an exception occurs. Default: false.

In the ``logger`` section, you may use the following options:

``level``
    The level at which you want to debug. May be ``debug``, ``info``,
    ``warning`` or ``error``. Default: ``warning``.

``path``
    Path to the log file, or a single dash (``-``) if logs should be
    written to the standard error stream. If given a path, it must be
    an absolute path. Default: ``-``.


Development
===========

AWStatic is hosted on GitHub at `<https://github.com/dbaty/AWStatic>`_.
Feel free to provide feedback and report issues or patches there.

AWStatic comes with automated tests that may be run with ``tox`` (that
uses both Python 2.7 and Python 3.2) or ``make test`` (that uses your
default ``python`` executable). The latter also opens
``awstatic/tests/js/tests.html`` that contains a test suite for the
JavaScript code.


Credits
=======

AWStatic includes the following libraries:

- `jQuery <http://jquery.com>`_, available under both the GPL and MIT license;

- `Flot <http://code.google.com/p/flot>`_, available under the MIT license;

- `Handlebars <http://handlebarsjs.com>`_, available under the MIT license.

It also includes an extract of `Bootstrap
<http://twitter.github.com/bootstrap>`_, available under the Apache
License v2.0.

The loading icon has been generated by `<http://www.ajaxload.info>`_.


License
=======

AWStatic is written by Damien Baty and is licensed under the 3-clause
BSD license, a copy of which is included in the source.

CHANGES.rst
List of releases of AWStatic
============================

AWStatic (0.1.0) - not yet released
-----------------------------------

First public release.

awstatic/tests/js/tests.js
$(document).ready(function () {
    // Test 'parse_querystring()'
    test('test_parse_querystring_empty', function() {
        same(awstatic.parse_querystring(''), {});
    });
    test('test_parse_querystring_basics', function() {
        same(awstatic.parse_querystring('foo=1&bar=2'),
             {'foo': '1', 'bar': '2'});
    });
    test('test_parse_querystring_trailing_ampersand', function() {
        same(awstatic.parse_querystring('foo=1&bar=2&'),
             {'foo': '1', 'bar': '2'});
    });
    test('test_parse_querystring_empty_value', function() {
        same(awstatic.parse_querystring('foo=1&bar='),
             {'foo': '1', 'bar': ''});
    });

    // Test 'to_querystring()'
    test('test_to_querystring_empty', function() {
        same(awstatic.to_querystring({}), '');
    });
    test('test_to_querystring_basics', function() {
        same(awstatic.to_querystring({'foo': '1', 'bar': '2'}), 'foo=1&bar=2');
    });
    test('test_to_querystring_empty_value', function() {
        same(awstatic.to_querystring({'foo': '1', 'bar': ''}), 'foo=1&bar=');
    });

    // Test 'get_period_label()'
    test('test_get_period_label_month_and_year', function() {
        same(awstatic.get_period_label('201201'), 'January 2012');
        same(awstatic.get_period_label('201202'), 'February 2012');
        same(awstatic.get_period_label('201203'), 'March 2012');
        same(awstatic.get_period_label('201204'), 'April 2012');
        same(awstatic.get_period_label('201205'), 'May 2012');
        same(awstatic.get_period_label('201206'), 'June 2012');
        same(awstatic.get_period_label('201207'), 'July 2012');
        same(awstatic.get_period_label('201208'), 'August 2012');
        same(awstatic.get_period_label('201209'), 'September 2012');
        same(awstatic.get_period_label('201210'), 'October 2012');
        same(awstatic.get_period_label('201211'), 'November 2012');
        same(awstatic.get_period_label('201212'), 'December 2012');
    });
    test('test_get_period_label_year_only', function() {
        same(awstatic.get_period_label('2012'), '2012');
    });

    // Test 'get_sorted_properties()'
    test('test_get_sorted_properties_basics', function() {
        same(awstatic.get_sorted_properties({'foo': 3, 'bar': 1, 'baz': 2}),
             [['bar', 1], ['baz', 2], ['foo', 3]]);
    });
    test('test_get_sorted_properties_empty_object', function() {
        same(awstatic.get_sorted_properties({}), []);
    });

    // Test 'is_leap_year()'
    test('test_is_leap_year', function() {
        ok(awstatic.is_leap_year(2012));
        ok(!awstatic.is_leap_year(2011));
        ok(awstatic.is_leap_year(1600));
        ok(!awstatic.is_leap_year(1700));
    });

    // Test 'get_month_ticks()'
    test('test_get_month_ticks', function() {
        same(awstatic.get_month_ticks('201201'),
             [[0, '1'], [1, '2'], [2, '3'], [3, '4'], [4, '5'], [5, '6'],
              [6, '7'], [7, '8'], [8, '9'], [9, '10'], [10, '11'], [11, '12'],
              [12, '13'], [13, '14'], [14, '15'], [15, '16'], [16, '17'],
              [17, '18'], [18, '19'], [19, '20'], [20, '21'], [21, '22'],
              [22, '23'], [23, '24'], [24, '25'], [25, '26'], [26, '27'],
              [27, '28'], [28, '29'], [29, '30'], [30, '31']]);
        same(awstatic.get_month_ticks('201202'),
             [[0, '1'], [1, '2'], [2, '3'], [3, '4'], [4, '5'], [5, '6'],
              [6, '7'], [7, '8'], [8, '9'], [9, '10'], [10, '11'], [11, '12'],
              [12, '13'], [13, '14'], [14, '15'], [15, '16'], [16, '17'],
              [17, '18'], [18, '19'], [19, '20'], [20, '21'], [21, '22'],
              [22, '23'], [23, '24'], [24, '25'], [25, '26'], [26, '27'],
              [27, '28'], [28, '29']]);
        same(awstatic.get_month_ticks('201102'),
             [[0, '1'], [1, '2'], [2, '3'], [3, '4'], [4, '5'], [5, '6'],
              [6, '7'], [7, '8'], [8, '9'], [9, '10'], [10, '11'], [11, '12'],
              [12, '13'], [13, '14'], [14, '15'], [15, '16'], [16, '17'],
              [17, '18'], [18, '19'], [19, '20'], [20, '21'], [21, '22'],
              [22, '23'], [23, '24'], [24, '25'], [25, '26'], [26, '27'],
              [27, '28']]);
    });

    // Test 'format_bandwidth()'
    test('test_format_bandwidth', function() {
        same(awstatic.format_bandwidth(undefined), '0');
        same(awstatic.format_bandwidth(0), '0');
        same(awstatic.format_bandwidth(1), '1 b');
        same(awstatic.format_bandwidth(1024), '1 Kb');
        same(awstatic.format_bandwidth(1132), '1.11 Kb');
        same(awstatic.format_bandwidth(2048 + 512), '2.5 Kb');
        same(awstatic.format_bandwidth(1024 * 1024), '1 Mb');
        same(awstatic.format_bandwidth(1024 * 1024 + 512 * 1024), '1.5 Mb');
        same(awstatic.format_bandwidth(1024 * 1024 * 1024), '1024 Mb');
    });
});

awstatic/tests/test_utils.py
from unittest import TestCase


class TestIsLeapYear(TestCase):

    def _call_fut(self, year):
        from awstatic.utils import is_leap_year
        return is_leap_year(year)

    def test_basics(self):
        self.assertTrue(self._call_fut(2012))
        self.assertFalse(self._call_fut(2011))
        self.assertTrue(self._call_fut(1600))
        self.assertFalse(self._call_fut(1700))


class TestNumberOfDays(TestCase):

    def _call_fut(self, yyyymm):
        from awstatic.utils import get_number_of_days
        return get_number_of_days(yyyymm)

    def test_basics(self):
        self.assertEqual(self._call_fut('201101'), 31)
        self.assertEqual(self._call_fut('201102'), 28)
        self.assertEqual(self._call_fut('201202'), 29)
        self.assertEqual(self._call_fut('201204'), 30)
        self.assertEqual(self._call_fut('201212'), 31)


class TestInterpolate(TestCase):

    def _call_fut(self, s, **bindings):
        from awstatic.utils import interpolate
        return interpolate(s, **bindings)

    def test_basics(self):
        self.assertEqual(self._call_fut('foo'), 'foo')
        self.assertEqual(self._call_fut('foo', foo='1'), 'foo')
        self.assertEqual(self._call_fut('${foo}'), '${foo}')
        self.assertEqual(self._call_fut('${foo}', bar='1'), '${foo}')
        self.assertEqual(self._call_fut('Value is: ${foo}', foo='1'),
                         'Value is: 1')


awstatic/tests/test_parser.py
from unittest import TestCase

import mock


class TestParser(TestCase):

    def _make_one(self):
        from awstatic.parser import Parser
        return Parser()

    @mock.patch('awstatic.parser.Parser.parse_file')
    def test_parse_dir(self, mock_parse_file):
        # Test that we read only the right files.
        import os
        here = os.path.dirname(__file__)
        in_dir = os.path.join(here, 'data', 'awstats')
        parser = self._make_one()
        parser.parse_dir('exemple.com', in_dir, 'awstats', 'txt')
        ap = lambda filename: os.path.join(in_dir, filename)
        expected = [((ap('awstats012012.exemple.com.txt'), '201201'), ),
                    ((ap('awstats022012.exemple.com.txt'), '201202'), ),
                    ((ap('awstats032012.exemple.com.txt'), '201203'), ),
                    ((ap('awstats042012.exemple.com.txt'), '201204'), ),
                    ((ap('awstats052012.exemple.com.txt'), '201205'), ),
                    ((ap('awstats062012.exemple.com.txt'), '201206'), )]
        self.assertEqual(mock_parse_file.call_args_list, expected)

    def test_parse_file_basics(self):
        import os
        here = os.path.dirname(__file__)
        path = os.path.join(here, 'data', 'awstats', 'basics.txt')
        parser = self._make_one()
        parser.parse_file(path, '201201')
        expected = {'GENERAL':
                        {'201201': {'TotalVisits': ['6'],
                                    'TotalUnique': ['6']}},
                    'VISITOR':
                        {'201201':
                             {'8.8.8.8.rev.sfr.net':
                                  {'host': '8.8.8.8.rev.sfr.net',
                                   'pages': '14',
                                   'hits': '38',
                                   'bandwidth': '1213892'},
                              'i04m-8-8-8-8.d4.club-internet.fr':
                                  {'host': 'i04m-8-8-8-8.d4.club-internet.fr',
                                   'pages': '2',
                                   'hits': '12',
                                   'bandwidth': '654661'},
                              }
                         }
                    }
        self.assertEqual(parser.data, expected)


awstatic/cli.py
"""Client entry point."""

from argparse import ArgumentParser
import logging
import os
import sys

from awstatic import __version__ as VERSION
from awstatic.compat import SafeConfigParser
from awstatic.reporter import Reporter


DEFAULT_CONFIG_FILE = 'awstatic.ini'
DEFAULT_LOG_LEVEL = 'warning'


def main():
    args = parse_args()
    config_file = args.config_file
    config = get_config(config_file)
    pdb_mode = config.pop('pdb')
    r = Reporter(**config)
    try:
        r.run()
    except:
        logging.exception('An unexpected error occurred (see traceback '
                          'below). Program aborted.')
        if pdb_mode:
            import pdb
            pdb.post_mortem()
        sys.exit(1)
    config['logger'].info('All reports have been successfully generated.')
    return 0


def parse_args():  # pragma: no coverage
    parser = ArgumentParser()
    add = parser.add_argument
    add('-v', '--version',
        action='version',
        version='%%(prog)s %s' % VERSION)
    add('config_file',
        metavar='CONFIG_FILE',
        nargs='?',
        default=DEFAULT_CONFIG_FILE,
        help='The configuration file to use. Default is '
             '"./%s".' % DEFAULT_CONFIG_FILE)
    return parser.parse_args()


def get_config(config_file):
    """Return a valid configuration (as a dictionary to be passed on
    to the constructor of the ``Reporter`` class) from the given
    configuration file.
    """
    config_parser = SafeConfigParser()
    if config_parser.read(config_file) != [config_file]:
        sys.exit('The configuration file at "%s" could '
                 'not be read.' % config_file)
    options = dict(config_parser.items('awstatic'))

    # Check unknown directives
    for key in options.keys():
        if key not in ('awstats_dir', 'file_prefix', 'file_suffix',
                       'sites', 'out_dir', 'pdb'):
            sys.exit('Unknown option in configuration file: "%s". '
                     'Program aborted.' % key)

    # Check required directives
    for opt in ('awstats_dir', 'out_dir', 'sites'):
        if opt not in options.keys():
            sys.exit('A required directive is missing from the '
                     'configuration file: "%s".' % opt)

    # Check directories
    awstats_dir = options['awstats_dir']
    awstats_dir = os.path.abspath(awstats_dir)
    if not os.path.isdir(awstats_dir):
        sys.exit('The value of "awstats_dir" ("%s") should be a valid '
                 'directory.' % awstats_dir)
    out_dir = options['out_dir']
    out_dir = os.path.abspath(out_dir)
    if not os.path.isdir(os.path.dirname(out_dir)):
        sys.exit('The parent of "out_dir" ("%s") must be an existing '
                 'directory.' % out_dir)
    if os.path.exists(out_dir) and not os.path.isdir(out_dir):
        sys.exit('The value of "out_dir" ("%s") should be a directory.' %
                 out_dir)

    # Prepare config dict and provide default values for optional
    # directives
    config = {'awstats_dir': awstats_dir,
              'out_dir': out_dir,
              'file_prefix': options.get('file_prefix', 'awstats'),
              'file_suffix': options.get('file_suffix', 'txt'),
              'sites': [],
              'pdb': options.get('pdb', '').lower() in ('1', 'true')}
    for id_url in options['sites'].split():
        error = False
        try:
            site_id, url = id_url.split('=', 1)
        except ValueError:
            error = True
        # Too few (error is True) or too many (url is a list)
        # '='-separated components
        if error or not isinstance(url, str):
            sys.exit('Wrong syntax for "sites".')
        config['sites'].append((site_id, url))

    config['logger'] = get_logger(dict(config_parser.items('logger')))
    return config


def get_logger(options):
    logger = logging.getLogger('AWStatic')
    level = options.get('level', DEFAULT_LOG_LEVEL).lower()
    level = {'debug': logging.DEBUG,
             'info': logging.INFO,
             'warning': logging.WARNING,
             'error': logging.ERROR}[level]
    logger.setLevel(level)
    path = options.get('path', '-')
    if path == '-':
        handler = logging.StreamHandler()
    else:
        # Error log file must be absolute, we do not want to guess.
        if os.path.abspath(path) != path:
            sys.exit('The path to the error file must be absolute.')
        handler = logging.FileHandler(path)
    formatter = logging.Formatter('%(asctime)s - %(levelname)s - %(message)s')
    handler.setFormatter(formatter)
    logger.addHandler(handler)
    return logger


if __name__ == '__main__':
    sys.exit(main())


awstatic/utils.py
import re


_DAYS_IN_MONTH = (31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31)
_INTERPOLATION = re.compile('\${(\w*)}')


def is_leap_year(year):
    """Return whether the given year is a leap year."""
    if year % 4:
        return False
    if year % 400 == 0:
        return True
    return year % 100 != 0


def get_number_of_days(yyyymm):
    """Return number of days of the given month."""
    month, year = map(int, (yyyymm[-2:], yyyymm[:-2]))
    n = _DAYS_IN_MONTH[month - 1]
    if month == 2 and is_leap_year(year):
        n += 1
    return n


def interpolate(s, **bindings):
    """Substitute ``${var}`` fragments in ``s`` by their value
    provided in ``bindings``.
    """
    if not bindings:
        return s
    def _sub(matchobj):
        var = matchobj.group(1)
        return bindings.get(var, matchobj.group(0))
    return _INTERPOLATION.sub(_sub, s)


awstatic/template/assets/js/ui.js
var awstatic = (function(window, $, undefined) {

"use strict";

/* ******************************************************************
** Utilities
** *********/
// Synchronous version of '$.getJSON()'.
function get_json(url) {
    return $.parseJSON(
        $.ajax({
            url: url,
            async: false,
            dataType: 'json',
            error: function (jqXHR, textStatus, errorThrown) {
                window.alert('Could not load "' + url + '": ' + errorThrown);
            }
        }).responseText);
}

// Parse the given query string and return an associative array with
// the key/values of the given hash. We use it here only to parse the
// hash part of the URL.
function parse_querystring(str) {
    if (!str) {
        return {};
    }
    var values = {};
    var groups = str.split('&');
    for (var i = 0; i < groups.length; i++) {
        var key_value = groups[i].split('=');
        if (!key_value[0].length) {
            continue;
        }
        values[key_value[0]] = key_value[1];
    }
    return values;
}

// Given an associate array, return the corresponding query string
// (without the leading '?').
function to_querystring(hash) {
    var values = [];
    for (var key in hash) {
        if (hash.hasOwnProperty(key)) {
            values.push(key + '=' + hash[key]);
        }
    }
    return values.join('&');
}

// Given a date formatted as 'YYYYMM' or 'YYYY', return a proper label
// such as 'month YYYY' or 'YYYY' (respectively).
var MONTHS = ['January', 'February', 'March', 'April', 'May', 'June',
    'July', 'August', 'September', 'October', 'November', 'December'];
function get_period_label(date) {
    if (date.length === 4) {
        return date;
    }
    return MONTHS[parseInt(date.slice(4), 10) - 1] +
        ' ' +
        date.slice(0, 4);
}

// Return the name and value (as an array of 2-element of arrays) of
// all properties of the given object, sorted by the property name.
function get_sorted_properties(obj) {
    var items = [];
    for (var prop in obj) {
        if (obj.hasOwnProperty(prop)) {
            items.push([prop, obj[prop]]);
        }
    }
    return items.sort();
}

// Return whether the given year is a leap year.
function is_leap_year(year) {
    if (year % 4 !== 0) {
        return false;
    }
    if (year % 400 === 0) {
        return true;
    }
    return (year % 100 !== 0);
}

// Return an array to be used as ticks in the graph, i.e. basically
// the list of days of the month.
// 'yyyymm' must be the month formatted as 'YYYYMM'. Duh.
var DAYS_IN_MONTH = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
function get_month_ticks(yyyymm) {
    var month = parseInt(yyyymm.slice(4), 10);
    var year = parseInt(yyyymm.slice(0, 4), 10);
    var n_days = DAYS_IN_MONTH[month - 1];
    if (month === 2 && is_leap_year(year)) {
        n_days += 1;
    }
    var ticks = [];
    for (var i = 0; i < n_days; i++) {
        ticks.push([i, String(1 + i)]);
    }
    return ticks;
}

// Given a numeric (or undefined value), return a string with the
// proper unit.
function format_bandwidth(bandwidth) {
    if (!bandwidth) {
        return '0';
    }
    var units = ['b', 'Kb', 'Mb'];
    var i = 1;
    for (; i < units.length; i++) {
        if (bandwidth < 1024) {
            break;
        }
        bandwidth = bandwidth / 1024;
    }
    // keep only 2 significant digits
    bandwidth = bandwidth.toFixed(2).toString();
    // remove trailing 0's.
    while (bandwidth.slice(bandwidth.length - 1) === '0') {
        bandwidth = bandwidth.slice(0, bandwidth.length - 1);
    }
    // remove dot if the 2 significant digits were 0's.
    if (bandwidth.slice(bandwidth.length - 1) === '.') {
        bandwidth = bandwidth.slice(0, bandwidth.length - 1);
    }
    return bandwidth + ' ' + units[i - 1];
}
/* *****************************************************************/


/* ******************************************************************
** Main 'UI' class
**
** The 'UI' class loads reports, generates diagrams and responds to
** user commands.
** *************************/
function UI(site, period, page) {
    this.init_templates();
    this.select_site(site, period);
    this.show_page(page || 'overview');
}

// Return the mode which we are in: 'month' or 'year'.
UI.prototype.get_period_mode = function() {
    if (this.period.length === 4) {
        return 'year';
    }
    return 'month';
};

// Compile templates.
UI.prototype.init_templates = function() {
    this._templates = {};
    var that = this;
    $('script[type="text/html"]').each(function(index, tag) {
        var template_id = tag.id.slice('tmpl-'.length);
        that._templates[template_id] = Handlebars.compile(
            $('#' + tag.id).html());
    });
};

// Render a template.
UI.prototype.render = function(template_id, data) {
    return this._templates[template_id](data);
};

// Select a site, load data, update reports and interface.
UI.prototype.select_site = function(new_site, new_period) {
    var old_site = this.site;
    var old_period = this.period;
    if ((new_site === old_site) && (new_period === old_period)) {
        return false;
    }
    $('#page-loading').show();
    if (new_site !== old_site) {
        this.site = new_site;
        this.data = get_json('data/' + this.site + '.json');
        this.url = this.data['url'];
        $('.domain-selector').find('.placeholder').html(new_site);
    }
    if (old_period !== new_period || old_period === undefined) {
        // If no particular period is provided, use the latest one.
        new_period = new_period || this.data['periods'][0];
        this.period = new_period;
        this.update_period_selector(this.data['periods']);
        var label = get_period_label(new_period);
        $('.period-selector').find('.placeholder').text(label);
    }
    this.update_reports();
    if (old_site === undefined) {
        // When the page is first loaded (and no site has been
        // selected yet), the header is empty. We want everything to
        // appear (text, content 'after:' and borders) at the same
        // time so the whole header has been set invisible. Here we
        // make it visible again.
        $('.header').children('li').toggleClass('invisible', false);
    }
    $('#page-loading').hide();
    this.update_hash();
};

// Select a period, update reports and interface.
UI.prototype.select_period = function(new_period) {
    if (new_period !== this.period) {
        this.select_site(this.site, new_period);
    }
};

// Update URL hash so that we can bookmark (deep-link) the page.
UI.prototype.update_hash = function() {
    window.location.hash = to_querystring({'site': this.site,
                                           'period': this.period,
                                           'page': this.page});
};

// Show one of the report pages.
UI.prototype.show_page = function(new_page) {
    if (this.page === new_page) {
        return;
    }
    $('#page-' + new_page).show();
    $('#nav-' + new_page).toggleClass('current', true);
    if (this.page !== undefined) {
        $('#page-' + this.page).hide();
        $('#nav-' + this.page).toggleClass('current', false);
    }
    this.page = new_page;
    this.update_hash();
};

// Update period dropdown selector.
UI.prototype.update_period_selector = function(periods, selected_label) {
    var hashes = [];
    for (var i = 0; i < periods.length; i++) {
        hashes.push({'key': periods[i], 'label': get_period_label(periods[i])});
    }
    $('#period-menu').html(this.render('period-menu', {'periods': hashes}));
};

// Update all reports.
UI.prototype.update_reports = function() {
    this.update_report_overview();
    this.update_report_top10();
    this.update_report_downloads();
    this.update_report_referrers();
    this.update_report_keywords();
    this.update_report_phrases();
};

// FIXME: review this. Probably rename.
// Return series for the selected period.
// - 'what' is the key in the JSON file. It may be either 'visits' or
//   'visits_by_day'. FIXME: not true anymore
// - 'mode' may be either 'month' or 'year'.
UI.prototype.get_series = function(what, mode) {
    var data = get_sorted_properties(this.data[what]);
    // 'idx' is the position of the varying component of the keys of
    // the 'data' object. The varying component is the day when in
    // 'month' mode, and the month when in 'year' mode.
    var idx = mode === 'year' ? 4 : 6;
    // 'expected_length' is the length of the keys that are
    // relevant. In 'year' mode, we look for YYYYMM keys (6 chars
    // long); in 'month' mode, we look for YYYYMMDD keys (8 chars
    // long).
    var expected_length = mode === 'year' ? 6 : 8;
    // We may not have data for the first days of the month (or the
    // first months of the year). Here we find out the first day (or
    // month) for which we have data.
    var i = 0;
    for (; i < data.length; i++) {
        if ((data[i][0].length === expected_length) &&
            (data[i][0].slice(0, idx) === this.period)) {
            break;
        }
    }
    var series = [];
    // We provide a value of 0 for missing days/months.
    var x = 0;
    var first = parseInt(data[i][0].slice(1 + idx), 10) - 1;
    for (; x < first; x++) {
        series.push([x, {}]);
    }
    // ... and add data that we do know.
    for (; i < data.length; i++) {
        if ((data[i][0].length === expected_length) &&
            (data[i][0].slice(0, idx) === this.period)) {
            series.push([x++, data[i][1]]);
        }
    }
    return series;
};

var YEAR_TICKS = [[0, 'jan'], [1, 'feb'], [2, 'mar'], [3, 'apr'], [4, 'may'],
    [5, 'jun'], [6, 'jul'], [7, 'aug'], [8, 'sep'], [9, 'oct'], [10, 'nov'],
    [11, 'dec']];
// Update overview page.
UI.prototype.update_report_overview = function() {
    var data = [];
    var options = {'series':
                   {'lines':
                    {'show': true,
                     'fill': 0.8},
                    'color': '#f6dd93',
                    'points': {'show': true}},
                   'yaxis': {'ticks': 2}}; // FIXME: really?
    // FIXME: add tooltips
    // (see http://people.iola.dk/olau/flot/examples/interacting.html)
    if (this.get_period_mode() === 'year') {
        data = this.get_series('overview', 'year');
        options['xaxis'] = {'ticks': YEAR_TICKS};
    }
    else { // mode === 'month'
        data = this.get_series('overview', 'month');
        options['xaxis'] = {'ticks': get_month_ticks(this.period)};
    }
    var series = [];
    for (var i = 0; i < data.length; i++) {
        series.push([data[i][0], data[i][1]['pages']]);
    }
    $.plot($("#overview-graph"), [series], options);

    // Update overview table.
    var table = [];
    for (var i = 0; i < data.length; i++) {
        var label = '';
        if (this.get_period_mode() === 'month') {
            label = String(1 + series[i][0]);
        } else {
            label = MONTHS[series[i][0]];
        }
        var bandwidth = format_bandwidth(data[i][1]['bandwidth']);
        table.push({'label': label,
                    'hits': data[i][1]['hits'],
                    'pages': data[i][1]['pages'],
                    'visits': data[i][1]['visits'],
                    'visitors': data[i][1]['visitors'],
                    'bandwidth': bandwidth});
    }
    var view = {'table': table};
    $('.overview').children('tbody').html(this.render('overview-table', view));
};

// Update "Top 10" report.
UI.prototype.update_report_top10 = function() {
    var pages = this.data['top10'][this.period] || [];
    var view = {'base_url': this.url,
                'pages': pages,
                'format_bandwidth': format_bandwidth};
    $('.top10').children('tbody').html(this.render('top10-table', view));
};

// Update "Downloads" report.
UI.prototype.update_report_downloads = function() {
    var files = this.data['downloads'][this.period] || [];
    var view = {'base_url': this.url,
                'files': files,
                'format_bandwidth': format_bandwidth};
    $('.downloads').children('tbody').html(
        this.render('downloads-table', view));
};

// Update "Referrers" report.
UI.prototype.update_report_referrers = function() {
    var referrers = this.data['referrers'][this.period] || [];
    var view = {'referrers': referrers};
    $('.referrers').children('tbody').html(
        this.render('referrers-table', view));
};

// Update "Search keywords" report.
UI.prototype.update_report_keywords = function() {
    var keywords = this.data['keywords'][this.period] || [];
    var view = {'keywords': keywords};
    $('.keywords').children('tbody').html(this.render('keywords-table', view));
};


// Update "Search phrases" report.
UI.prototype.update_report_phrases = function() {
    var phrases = this.data['phrases'][this.period] || [];
    var view = {'phrases': phrases};
    $('.phrases').children('tbody').html(this.render('phrases-table', view));
};


var ui = undefined; // will be set in 'init_ui()'

// Initialize the user interface (to be called when the document is
// ready): this sets up dropdown menus, generates graphs, etc.
function init_ui() {
    $('.dropdown-toggle').dropdown();
    var sites = get_json('data/sites.json');
    var qs = parse_querystring(window.location.hash);
    var site = qs.site;
    if (!site) {
        site = sites[0];
    }
    // Set 'window.ui'
    // 'qs.period' and 'qs.page' may be undefined but the constructor
    // can deal with that.
    window.ui = ui = new UI(site, qs.period, qs.page);
    $('#site-menu').html(ui.render('site-menu', {'sites': sites}));
}

// public symbols of the module
return {
    format_bandwidth: format_bandwidth,
    get_month_ticks: get_month_ticks,
    get_period_label: get_period_label,
    get_sorted_properties: get_sorted_properties,
    init_ui: init_ui,
    is_leap_year: is_leap_year,
    parse_querystring: parse_querystring,
    to_querystring: to_querystring
};

// end of module
})(window, jQuery);